compiler/rustc_arena/src/lib.rs

```diff
@@ -136,51 +136,39 @@ where
 impl<T, const N: usize> IterExt<T> for std::array::IntoIter<T, N> {
     #[inline]
     fn alloc_from_iter(self, arena: &TypedArena<T>) -> &mut [T] {
-        let len = self.len();
-        if len == 0 {
-            return &mut [];
-        }
-        // Move the content to the arena by copying and then forgetting it
-        unsafe {
-            let start_ptr = arena.alloc_raw_slice(len);
-            self.as_slice().as_ptr().copy_to_nonoverlapping(start_ptr, len);
-            mem::forget(self);
-            slice::from_raw_parts_mut(start_ptr, len)
-        }
+        alloc_from_iter(arena, self.len(), self.as_slice().as_ptr(), || mem::forget(self))
```

Comment:
I am not sure whether we're not using this or if LLVM optimizations are just making this okay, but this is surely UB, right? The raw pointer passed into the alloc_from_iter function points at self on the stack (presumably) and then we move that into the closure, invalidating the pointer.

(I should also note that alloc_from_iter is clearly unsafe)

Reply:
Wait, I don't quite understand this part, not sure if it is UB. But that closure is run at the end right? Shouldn't the closure `self` still points at the same thing?

Reply:
When the closure runs doesn't matter; it captures `self` by value (moving it into the closure environment).

Reply:
Oh, then it is wrong. Should I close the PR? I didn't noticed this, I am still kinda confused. I thought it is safe since the closure runs at a certain point. Or maybe should I make a noop function for this?

Reply:
I don't have the time right now to explain further, but I'm guessing that you could ask on Zulip for clarification. It *is* subtle, and I mainly noticed because I was surprised we didn't need move annotations on the closures. (I could also be wrong; but I don't see any way for this to not be a bug given the signature on mem::forget).

We do need to forget the array as we've moved its elements into the arena. We could do so with ManuallyDrop, I suppose, but depending on LLVM's optimizations that might have its own overheads. I personally don't think the duplication of the 3 or so lines is a big problem here.

Reply:
Does that mean `alloc_from_iter` is unsound, since it allows UB without unsafe? It seems weird for a safe function to take a raw pointer.

Reply:
@jyn514 It is, but it is private so it is perhaps forgivable. (I would still mark it as unsafe though to prevent issues like this one.)

Reply:
Oh oh I see, alloc_from_iter is a newly added function that should have been marked unsafe but wasn't.